src/flow.rs

```diff
@@ -0,0 +1,192 @@
+};
+
+type Flow = hashbrown::HashMap<usize, usize>;
+
+/// Checks the definition of causal flow.
+///
+/// 1. i -> f(i)
+/// 2. j in neighbors(f(i)) => i == j or i -> j
+/// 3. i in neighbors(f(i))
+fn check_definition(f: &Flow, layer: &Layer, g: &Graph) -> anyhow::Result<()> {
+    for (&i, &fi) in f.iter() {
+        if layer[i] <= layer[fi] {
+            let err = anyhow::anyhow!("layer check failed").context(format!("must be {i} -> {fi}"));
+            return Err(err);
+        }
+        for &j in &g[fi] {
+            if i != j && layer[i] <= layer[j] {
+                let err = anyhow::anyhow!("layer check failed")
+                    .context(format!("neither {i} == {j} nor {i} -> {j}"));
+                return Err(err);
+            }
+        }
+        if !(g[fi].contains(&i) && g[i].contains(&fi)) {
+            let err = anyhow::anyhow!("graph check failed")
+                .context(format!("{i} and {fi} not connected"));
+            return Err(err);
+        }
+    }
+    Ok(())
+}
+
+/// Finds the maximally-delayed causal flow.
+///
+/// # Arguments
+///
+/// - `g`: The adjacency list of the graph. Must be undirected and without self-loops.
+/// - `iset`: The set of initial nodes.
+/// - `oset`: The set of output nodes.
+///
+/// # Note
+///
+/// - Node indices are assumed to be `0..g.len()`.
+/// - Arguments are **NOT** verified.
+#[pyfunction]
+pub fn find(g: Graph, iset: Nodes, mut oset: Nodes) -> Option<(Flow, Layer)> {
+    log::debug!("flow::find");
+    validate::check_graph(&g, &iset, &oset).unwrap();
+    let n = g.len();
+    let vset = (0..n).collect::<Nodes>();
+    let mut cset = &oset - &iset;
+    let icset = &vset - &iset;
+    let ocset = &vset - &oset;
+    let oset_orig = oset.clone();
+    let mut f = Flow::with_capacity(ocset.len());
+    let mut layer = vec![0_usize; n];
+    // check[v] = g[v] & (vset - oset)
+    let mut check = g.iter().map(|x| x & &ocset).collect::<Vec<_>>();
+    let mut oset_work = Nodes::new();
+    let mut cset_work = Nodes::new();
+    for l in 1_usize.. {
+        log::debug!("=====layer {l}=====");
+        oset_work.clear();
+        cset_work.clear();
+        for &v in &cset {
+            let checkv = &check[v];
+            if checkv.len() != 1 {
+                continue;
+            }
+            let u = *checkv.iter().next().expect("one element here");
+            log::debug!("f({u}) = {v}");
+            f.insert(u, v);
+            log::debug!("layer({u}) = {l}");
+            layer[u] = l;
+            oset_work.insert(u);
+            cset_work.insert(v);
+        }
+        if oset_work.is_empty() {
+            break;
+        }
+        // For all u check[u] -= oset_work
+        for &v in &oset_work {
+            g[v].iter().for_each(|&u| {
+                check[u].remove(&v);
+            });
+        }
+        oset.union_with(&oset_work);
+        cset.difference_with(&cset_work);
+        cset.union_with(oset_work.intersection(&icset));
```

Comment:
If I understand correctly, `union_with` is just `extend`.

```suggestion
        oset.extend(&oset_work);
        cset.difference_with(&cset_work);
        cset.extend(oset_work.intersection(&icset));
```

Reply:
I'm using the name `union_with` to clarify what I definitely want to do.

Reply:
I would prefer to use standard names whenever possible. Moreover, the implementation of union_with is not just an alias for extend; it contains clone operations that we usually want to eliminate when possible in Rust.

Reply:
Fixed in 1166afde2d094c116bcdd982196b5c38921e65e2.